tests/test_cfg/test_z3_constraints.py

```diff
@@ -154,7 +154,59 @@ def func(x: int, y: int) -> None:
     cfg = _create_cfg(src, "func")
     x = z3.Int("x")
     y = z3.Int("y")
-    expected_while_true_path = [{x > 5, y > 10}, {x > 5, y > 10, x + y > 15}]
+    expected_while_true_path = [
+        {x > 5, y > 10},
+        {x > 5, y > 10, x + y > 15},
+        {},
```

Comment:
It seems quite strange for this to be here. What edge is this supposed to correspond to?

All edges should, at a minimum, have the precondition constraints. So if you're getting an empty set something seems off. (Note: `{}` is an empty dict literal, which is even more suspicious.)

Reply:
Hi Professor, I think this is the expected behavior. This edge corresponds to the highlighted one pointing from loop body to condition. Since both `x` and `y` are reassigned in the loop body, all z3 constraints that involves `x` and `y`, including precondition, are removed.

<img width="209" alt="image" src="https://github.com/user-attachments/assets/dd5d09d3-578e-4606-bdd9-de968a7f8f96">

Reply:
Okay that makes sense, thanks. The value is fine, but see my other comment about `{}` not being the appropriate data type. It's strange that the tests didn't pick up on that, you should investigate why not.

Reply:
Hi Professor, thanks for the reminder. Even though we do expect an empty set here, but `{}` is an empty dictionary. I will definitely look into what's going on here.

Reply:
Hi Professor, I’ve changed the empty dictionary to `set()`. The reason the tests passed previously is that they iterate through and compare each z3 constraint using `zip(actual, expected)`. When the collection is empty—whether it’s a set or a dictionary—this iteration doesn’t execute. This behavior shouldn’t cause any issues since we’ve already verified that the lengths of expected and actual are equal beforehand.

Reply:
okay great 👍